### PR TITLE
Add validation for all container at once, refactoring and other changes

### PR DESCRIPTION
**Pull request by vmahedia**
This changes are one of the many incremental that need to happen in order to make the cli user experience smoother. With this change, all the container configs with be validated and all deployment checks will be run before we deploy single container out of many, earlier we did validation/deploy one by one where customers complained that they wanted to know the validation error for all containers before any of them is deployed.

### CODE REVIEW

File: cli/roger_push.py [hunk 1]
```diff
@@ -384,102 +440,106 @@ def main(self, settings, appConfig, frameworkObject, hooksObj, args):
...
                                 execution_result = 'FAILURE'
                                 self.outcome = 0
 
-                            time_take_milliseonds = ((datetime.now() - function_execution_start_time).total_seconds() * 1000)
+                            time_taken = (datetime.now() - function_execution_start_time).total_seconds()
                             for task_id in container_task_id:
-                                input_metric = "roger-tools.rogeros_tools_exec_time," + "app_name=" + str(args.app_name) + ",event=push" + ",container_name=" + str(container_name) + ",identifier=" + str(self.identifier) + ",outcome=" + str(execution_result) + ",response_code=" + str(status_code) + ",config_name=" + str(config_name) + ",env=" + str(environment) + ",user=" + str(settingObj.getUser()) + ",task_id=" + str(task_id) + ",tools_version=" + str(tools_version_value) + ",image_tag=" + str(image_tag_value)
-                                tup = (input_metric, time_take_milliseonds)
-                                self.statsd_push_list.append(tup)
+                                input_metric = "roger-tools.rogeros_tools_exec_time" + \
```
Comment (jhulten) [inline cli/roger_push.py]: This will need to change for datadog

File: cli/roger_push.py [hunk 2]
```diff
@@ -174,9 +179,13 @@ def repo_relative_path(self, appConfig, args, repo, path):
             return "{0}/{1}/{2}/{3}".format(os.environ.get('PWD', ''),
                                             args.directory, repo_name, path)
 
+    def getContainerName(self, container):
+         return str(container.keys()[0]) if type(container) == dict else container
```
Comment (jhulten) [inline cli/roger_push.py]: Are we leaving ourselves open to issues by assuming a string if it isn't a dict?

File: cli/roger_push.py [hunk 3]
```diff
@@ -384,102 +440,106 @@ def main(self, settings, appConfig, frameworkObject, hooksObj, args):
...
 
                         except (Exception) as e:
-                            print("The following error occurred: %s" %e, file=sys.stderr)
+                            print("The following error occurred: %s" % e, file=sys.stderr)
                             raise
 
             hooksObj.statsd_message_list = self.statsd_message_list
             hookname = "post_push"
-            hook_input_metric = "roger-tools.rogeros_tools_exec_time," + "event=" + hookname + ",app_name=" + str(args.app_name) + ",identifier=" + str(self.identifier) + ",config_name=" + str(config_name) + ",env=" + str(environment) + ",user=" + str(settingObj.getUser())
+            hook_input_metric = "roger-tools.rogeros_tools_exec_time," + "event=" + hookname + \
```
Comment (jhulten) [inline cli/roger_push.py]: Likewise, this is a potential place for datadog conversion

File: cli/roger_push.py [hunk 4]
```diff
@@ -215,22 +224,32 @@ def main(self, settings, appConfig, frameworkObject, hooksObj, args):
...
+            # ----------------------------------------------
             environmentObj = roger_env['environments'][environment]
             common_repo = config.get('repo', '')
+
+            # ----------------------------------------------
+            # GetContainersForApp(app)
+            # ----------------------------------------------
             app_name = args.app_name
             container_list = []
+            # todo (vmahedia): What does ':' signify? Put explanation.
```
Comment (Hosweay) [inline cli/roger_push.py]: The : and the , are used when deploying multiple. The : is supposed to signify multiple containers under a common application name ala feedstore, https://marathon-prod-roger.dal.moz.com/ui/#/group/%2Fmoz-kwe%2Ffeedstore. The , is supposed to be multiple different applications I believe.

File: cli/marathon.py [hunk 5]
```diff
@@ -202,6 +202,10 @@ def getCurrentImageVersion(self, roger_env, environment, application):
         data = self.get(roger_env, environment)
         self.fetchUserPass(environment)
         for app in data['apps']:
+            # (vmahedia) todo: there's a YOOGE bug here, this just picks the first image it founds
+            # with an app name in it, this is wrong, cli should either be very explicity or not try
+            # to guess at all. We need to change this bigly.
```
Comment (asmidutz) [inline cli/marathon.py]: Support the use of "bigly"

**Pull request merged**